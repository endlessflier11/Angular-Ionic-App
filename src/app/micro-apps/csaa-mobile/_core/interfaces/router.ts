import { Route, UrlTree } from '@angular/router';
import { NavigationOptions } from '@ionic/angular/providers/nav-controller';
import { Observable } from 'rxjs';

interface AppRoute extends Route {
  name?: string;
  isGroupIndex?: boolean; // If true, it means route is the index of parent. Then it inherits all parent properties
}

export type AppRoutes = AppRoute[];

export interface RouteParam {
  [key: string]: string;
}

export interface NamedRouteConfig {
  name: string;
  path: string;
  children?: NamedRouteConfig[];
}

export interface NamedRoute {
  name: string;
  path: string;
  relativePath: string;
}

export interface CanComponentDeactivate {
  canDeactivate: () =>
    | Observable<boolean | UrlTree>
    | Promise<boolean | UrlTree>
    | boolean
    | UrlTree;
}

export interface CsaaRouteStackItem {
  name: string;
  params?: RouteParam;
  options?: NavigationOptions;
}
