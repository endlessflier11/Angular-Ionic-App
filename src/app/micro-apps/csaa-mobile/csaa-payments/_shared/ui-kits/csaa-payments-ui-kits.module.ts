import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { CommonModule } from '@angular/common';
import { CsaaPaymentHistoryCardComponent } from './csaa-payment-history-card/csaa-payment-history-card.component';
import { UiKitsModule } from '../../../_core/ui-kits/ui-kits.module';
import { CsaaPaymentHistorySelectCardComponent } from './csaa-payment-history-select-card/csaa-payment-history-select-card.component';
import { CsaaPaymentSucceededCardComponent } from './csaa-payment-succeeded-card/csaa-payment-succeeded-card.component';
import { CsaaMakePaymentCardComponent } from './csaa-make-payment-card/csaa-make-payment-card.component';
import { CsaaPayAllPoliciesCardComponent } from './csaa-pay-all-policies-card/csaa-pay-all-policies-card.component';
import { PaymentTermsConditionsComponent } from './payment-terms-conditions/payment-terms-conditions.component';
import { AutopayTermsConditionsComponent } from './autopay-terms-conditions/autopay-terms-conditions.component';
import { CsaaBillsCardComponent } from './csaa-bills-card/csaa-bills-card.component';
import { AutopayDiscountCardComponent } from './autopay-discount-card/autopay-discount-card.component';

const SHARED_COMPONENTS = [
  CsaaPaymentHistoryCardComponent,
  CsaaPaymentHistorySelectCardComponent,
  CsaaMakePaymentCardComponent,
  CsaaPaymentSucceededCardComponent,
  CsaaPayAllPoliciesCardComponent,
  PaymentTermsConditionsComponent,
  AutopayTermsConditionsComponent,
  CsaaBillsCardComponent,
  AutopayDiscountCardComponent,
];

@NgModule({
  declarations: [...SHARED_COMPONENTS],
  imports: [CommonModule, IonicModule, UiKitsModule, IonicModule],
  exports: [...SHARED_COMPONENTS],
})
export class CsaaPaymentsUiKitsModule {}
