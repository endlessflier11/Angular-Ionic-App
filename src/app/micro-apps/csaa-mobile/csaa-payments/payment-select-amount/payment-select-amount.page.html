<ion-header [ngClass]="['csaa', currentTheme]">
  <ion-toolbar>
    <ion-buttons slot="start">
      <ion-button (click)="cancel()">Cancel</ion-button>
    </ion-buttons>
    <ion-title>Amount</ion-title>
    <ion-buttons slot="end">
      <ion-button (click)="selectAmount()">Done</ion-button>
    </ion-buttons>
  </ion-toolbar>
</ion-header>
<ion-content [ngClass]="['csaa', currentTheme]" class="csaa-item-list">
  <h3 class="ion-text-center">Select an amount</h3>

  <form [formGroup]="amountForm">
    <csaa-skeleton-card
      *ngIf="!paymentData"
      [showHeader]="false"
      [showButton]="false"
    ></csaa-skeleton-card>
    <ion-list lines="full" *ngIf="!!paymentData">
      <ion-item
        id="item-minimum"
        (click)="selectAmountType('minimum')"
        [class.active]="paymentData?.type === 'minimum'"
      >
        <ion-label>Minimum Due</ion-label>
        <span slot="end">
          {{ paymentData?.minimumAmount | currency: 'USD':'symbol':'1.2-2' }}
        </span>
      </ion-item>

      <ion-item
        id="item-remaining"
        (click)="selectAmountType('remaining')"
        [class.active]="paymentData?.type === 'remaining'"
      >
        <ion-label>Remaining Premium</ion-label>
        <span item-end>
          {{ paymentData?.remainingPremium | currency: 'USD':'symbol':'1.2-2' }}
        </span>
      </ion-item>

      <ion-item
        *ngIf="!paymentData?.allPolicies"
        id="item-other"
        (click)="selectAmountType('other'); otherAmountInput.setFocus();"
        [class.active]="paymentData?.type === 'other'"
      >
        <ion-label>Other Amount</ion-label>
        <ion-input
          formControlName="otherAmount"
          type="{{isIos?'text':'tel'}}"
          class="ion-text-right"
          [brmasker]="otherAmountMaskConfig"
          #otherAmountInput
        >
        </ion-input>
      </ion-item>
    </ion-list>
  </form>
</ion-content>
