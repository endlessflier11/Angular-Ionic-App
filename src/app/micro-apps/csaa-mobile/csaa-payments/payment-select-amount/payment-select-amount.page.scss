ion-list {
  padding-top: 0;
  padding-bottom: 0;
}
h3 {
  margin-bottom: 20px;
}
ion-content {
  --background: white;
}
ion-item {
  --highlight-color-valid: var(--ion-color-primary);
  &:first-child {
    border-top: 1px solid #c8c7cc;
  }

  ion-input {
    visibility: hidden;
  }

  &.active {
    --border-width: 2px;
    --border-color: var(--ion-color-primary);
    span {
      color: var(--ion-color-primary);
    }

    ion-input {
      visibility: initial;
    }
  }
}
