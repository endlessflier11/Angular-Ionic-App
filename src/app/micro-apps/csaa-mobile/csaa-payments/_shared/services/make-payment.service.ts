import { Injectable } from '@angular/core';
import { NavigationEnd, Router } from '@angular/router';
import { filter } from 'rxjs/operators';
import { Subscription } from 'rxjs';

import {
  PaymentAccount,
  PaymentResult,
  UpcomingPayment,
  UpcomingPaymentModel,
} from '../../../_core/interfaces';
import { RouterService } from '../../../_core/services';
import { withErrorReporter } from '../../../_core/helpers';
import { CsaaCommonModule } from '../../../csaa-core/csaa-common.module';
import { format } from 'date-fns';

@Injectable({
  providedIn: CsaaCommonModule,
})
export class MakePaymentService {
  private resetListenerSub: Subscription;
  private upcomingPayments: UpcomingPayment[];
  private activePayment: UpcomingPayment;
  private activePaymentAccount: PaymentAccount;
  private paymentResults: PaymentResult = {};

  private returnPathFromAmountMethodPages: { path: string; params?: any };

  public get hasPayment(): boolean {
    return !!this.activePayment;
  }

  constructor(private readonly router: Router, private routerService: RouterService) {}

  startResetListener(): void {
    if (this.resetListenerSub) {return;}
    this.resetListenerSub = this.router.events
      .pipe(filter((e) => e instanceof NavigationEnd))
      .subscribe(
        withErrorReporter((e: NavigationEnd) => {
          if (!e.url.startsWith(this.routerService.fullPath('csaa.payment.index'))) {
            this.reset();
            this.stopResetListener();
          }
        })
      );
  }

  stopResetListener(): void {
    if (!this.resetListenerSub) {return;}
    this.resetListenerSub.unsubscribe();
    this.resetListenerSub = undefined;
  }

  public reset(): void {
    this.setActivePayment(null);
    this.selectPaymentAccount(null);
    this.setReturnPathFromAmountMethodPages(null);
    this.setPaymentResult({});
  }

  public setActivePayment(payment: UpcomingPayment): void {
    this.activePayment = null;
    if (!!payment) {
      this.activePayment = new UpcomingPaymentModel({ ...payment });
    }
    this.startResetListener();
  }

  public getActivePayment() {
    return !!this.activePayment ? new UpcomingPaymentModel({ ...this.activePayment }) : null;
  }

  public updatePaymentAmount(value: number) {
    this.activePayment.otherAmount = value;
  }

  public getPaymentAmount(): number {
    const { amount, otherAmount } = this.activePayment;
    return otherAmount || amount;
  }

  public setUpcomingPayments(payments: UpcomingPayment[]): void {
    this.upcomingPayments = payments;
    this.startResetListener();
  }

  public getUpcomingPayments(): UpcomingPayment[] {
    return this.upcomingPayments;
  }

  public selectPaymentAccount(account: PaymentAccount) {
    this.activePaymentAccount = account;
  }

  public pullSelectedPaymentAccount(): PaymentAccount {
    const activePaymentAccount = this.activePaymentAccount;
    this.activePaymentAccount = null;
    return activePaymentAccount;
  }

  public setReturnPathFromAmountMethodPages(path: string, params = null) {
    this.returnPathFromAmountMethodPages = { path, params };
  }

  public getReturnPathFromAmountMethodPages() {
    return this.returnPathFromAmountMethodPages || { path: 'csaa.payment.index' };
  }

  public setPaymentResult(results: PaymentResult) {
    this.paymentResults = { ...results };
  }

  public pullPaymentResults() {
    const results = this.paymentResults;
    this.paymentResults = null;
    return { ...results };
  }

  public updatePaymentResult(
    policyNumber: string,
    status: string,
    receiptNumber: string,
    paymentAccount: PaymentAccount,
    amount: number
  ) {
    this.paymentResults[policyNumber] = {
      result: status === 'SUCC',
      receiptNumber,
      paymentAccount,
      amount,
      date: format(new Date(Date.now()), 'yyyy-MM-dd'),
    };
  }

  public getPaymentResult(policyNumber: string): boolean {
    return !!this.paymentResults[policyNumber];
  }
}
