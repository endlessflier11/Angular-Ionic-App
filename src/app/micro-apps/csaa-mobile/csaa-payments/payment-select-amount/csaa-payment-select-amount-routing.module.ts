import { NgModule } from '@angular/core';
import { PaymentSelectAmountPage } from './payment-select-amount.page';
import { RouterModule } from '@angular/router';
import { ReactiveFormsModule } from '@angular/forms';
import { AppRoutes } from '../../_core/interfaces';

const routes: AppRoutes = [
  {
    path: '',
    isGroupIndex: true,
    component: PaymentSelectAmountPage,
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes), ReactiveFormsModule],
  exports: [RouterModule],
})
export class CsaaPaymentSelectAmountRoutingModule {}
