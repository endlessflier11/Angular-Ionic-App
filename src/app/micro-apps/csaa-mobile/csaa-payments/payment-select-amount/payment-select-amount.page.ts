import { Component, OnInit } from '@angular/core';
import { MakePaymentService } from '../_shared/services/make-payment.service';
import { FormBuilder, FormGroup } from '@angular/forms';
import { AlertController, Platform } from '@ionic/angular';
import { noop } from 'rxjs';
import {
  IonViewWillEnter,
  UpcomingPayment,
  PolicyType,
  EventName,
  Category,
  EventType,
} from '../../_core/interfaces';
import { ConfigService, RouterService, AnalyticsService } from '../../_core/services';
import { PaymentState } from '../../_core/store/states/payment.state';
import { Store } from '@ngxs/store';

@Component({
  selector: 'app-csaa-payment-select-amount',
  templateUrl: './payment-select-amount.page.html',
  styleUrls: ['./payment-select-amount.page.scss'],
})
export class PaymentSelectAmountPage implements OnInit, IonViewWillEnter {
  currentTheme: string;

  amountForm: FormGroup;
  paymentData: UpcomingPayment;
  payments: UpcomingPayment[];
  isIos: boolean;

  public get otherAmountMaskConfig() {
    return {
      money: true,
      numberAndTousand: true,
      thousand: ',',
      decimal: 2,
      decimalCaracter: '.',
    };
  }

  constructor(
    private readonly configService: ConfigService,
    private readonly routerService: RouterService,
    private readonly makePaymentService: MakePaymentService,
    private readonly formBuilder: FormBuilder,
    private readonly platform: Platform,
    private readonly alertCtrl: AlertController,
    private readonly analyticsService: AnalyticsService,
    private readonly store: Store
  ) {
    this.currentTheme = this.configService.getTheme();

    if (!makePaymentService.hasPayment) {
      this.navigateBack();
    }
    this.amountForm = this.formBuilder.group({
      otherAmount: [''],
    });
  }

  ngOnInit() {}

  ionViewWillEnter(): void {
    this.payments = this.store.selectSnapshot(PaymentState.upcomingPayments);

    this.paymentData = this.makePaymentService.getActivePayment();
    if (this.paymentData && !this.paymentData.allPolicies) {
      const { otherAmount } = this.paymentData;
      this.amountForm.get('otherAmount').setValue(otherAmount);
    }

    if (!this.paymentData) {
      this.navigateBack();
    }

    this.isIos = this.platform.is('ios');
  }

  get otherAmount() {
    const { otherAmount } = this.amountForm.controls;
    return parseFloat(otherAmount && otherAmount.value ? otherAmount.value.replace(/,/, '') : 0);
  }

  async selectAmount() {
    this.paymentData.otherAmount = this.otherAmount;
    const isCustomAmount = this.paymentData.otherAmount !== 0;

    if (isCustomAmount && this.paymentData.otherAmount < this.paymentData.minimumAmount) {
      const alert = await this.alertCtrl.create({
        header: 'You are paying less than the minimum due.',
        message: 'Paying less than the minimum may result in policy cancelling.',
        buttons: [
          {
            text: 'Cancel',
            role: 'cancel',
            handler: () => {
              this.trackAmountSelectedEvent(false, false);
              alert.dismiss(false);
              return false;
            },
          },
          {
            text: 'Ok',
            handler: () => {
              this.trackAmountSelectedEvent(true, false);
              alert.dismiss(false);
              this.makePaymentService.setActivePayment(this.paymentData);
              this.navigateBack();
              return false;
            },
          },
        ],
      });
      alert.present();
    } else if (isCustomAmount && this.paymentData.otherAmount > this.paymentData.remainingPremium) {
      const alert = await this.alertCtrl.create({
        header: 'Payment error',
        message: 'You cannot pay more than the remaining premium.',
        buttons: [
          {
            text: 'Close',
            role: 'cancel',
            handler: () => {
              alert.dismiss(false);
              return false;
            },
          },
        ],
      });
      alert.present();
    } else {
      this.makePaymentService.setActivePayment(this.paymentData);
      this.routerService.back();
    }
  }

  private trackAmountSelectedEvent(confirm: boolean, coversMinimumAmount: boolean) {
    const { payments, paymentData } = this;
    const eventProperties = {
      event_type: EventType.OPTION_SELECTED,
      selection: confirm ? 'Confirm' : 'Decline',
      min_amt_paid: coversMinimumAmount,
      policies: (paymentData.allPolicies
        ? payments
        : payments.filter((p) => p.policyNumber === paymentData.policyNumber)
      ).map((p) => ({
        policy_number: p.policyNumber,
        policy_type: PolicyType[p.policyType],
        policy_state: p.policyRiskState,
      })),
    };
    this.analyticsService.trackEvent(
      EventName.PAYMENT_AMOUNT_CONFIRMED,
      Category.payments,
      eventProperties
    );
  }

  selectAmountType(type) {
    this.paymentData.type = type;

    if (type !== 'other') {
      this.paymentData.otherAmount = 0;
      this.amountForm.controls.otherAmount.setValue('');
    }
  }

  cancel() {
    this.analyticsService.trackEvent(EventName.PAYMENT_AMOUNT_CANCELLED, Category.payments);
    this.navigateBack();
  }

  private navigateBack() {
    const { path, params } = this.makePaymentService.getReturnPathFromAmountMethodPages();
    this.routerService.navigateBack(path, params).then(noop);
  }
}
