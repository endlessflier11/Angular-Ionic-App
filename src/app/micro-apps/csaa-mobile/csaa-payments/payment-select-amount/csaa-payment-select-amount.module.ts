import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CsaaPaymentSelectAmountRoutingModule } from './csaa-payment-select-amount-routing.module';
import { PaymentSelectAmountPage } from './payment-select-amount.page';
import { IonicModule } from '@ionic/angular';
import { UiKitsModule } from '../../_core/ui-kits/ui-kits.module';
import { ReactiveFormsModule } from '@angular/forms';
import { BrMaskerModule } from 'br-mask';

@NgModule({
  declarations: [PaymentSelectAmountPage],
  imports: [
    CommonModule,
    ReactiveFormsModule,
    CsaaPaymentSelectAmountRoutingModule,
    IonicModule,
    UiKitsModule,
    BrMaskerModule,
  ],
})
export class CsaaPaymentSelectAmountModule {}
