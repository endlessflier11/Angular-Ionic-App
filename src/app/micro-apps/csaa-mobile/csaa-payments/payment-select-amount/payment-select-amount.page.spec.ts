import { ComponentFixture, TestBed } from '@angular/core/testing';
import { AlertController, IonicModule } from '@ionic/angular';

import { PaymentSelectAmountPage } from './payment-select-amount.page';
import { AlertControllerMock, By, PageTestingModule } from '@app/testing';
import { UiKitsModule } from '../../_core/ui-kits/ui-kits.module';
import { MakePaymentService } from '../_shared/services/make-payment.service';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { PaymentService } from '../../_core/services/payment.service';
import { of } from 'rxjs';
import {
  Category,
  EventName,
  PaymentType,
  PolicyType,
  UpcomingPayment,
} from '../../_core/interfaces';
import { BrMaskDirective } from 'br-mask';
import DateHelper from '../../_core/shared/date.helper';
import { AnalyticsService, RouterService } from '../../_core/services';

const mockPaymentData = {
  vehicles: ['Lexus NX', 'BMW X1'],
  subtitle: 'test',
  autoPay: true,
  minimumAmount: 20,
  remainingPremium: 10000,
  otherAmount: 1000,
  type: PaymentType.other,
  nextDueDate: DateHelper.toDate('2040-08-19'),
  nextAmount: 30,
  allPolicies: false,
  policyNumber: 'AZSS294857564',
  policyType: PolicyType.Auto,
  dueDate: DateHelper.toDate('2018-08-10'),
  isPastDue: true,
  isPaymentDue: true,
  amount: 260.75,
  policyRiskState: 'test',
  autopayEnrollment: null,
};

describe('PaymentSelectAmountPage', () => {
  let analyticService;
  let routerService;
  let component: PaymentSelectAmountPage;
  let fixture: ComponentFixture<PaymentSelectAmountPage>;
  let makePaymentService: MakePaymentService;
  let alertControllerMock: AlertControllerMock;

  beforeEach(async () => {
    try {
      await TestBed.configureTestingModule({
        declarations: [PaymentSelectAmountPage, BrMaskDirective],
        imports: [
          RouterTestingModule,
          PageTestingModule.withConfig({
            providesStorage: true,
            providesConfig: true,
            providesRouter: true,
            providesAlert: true,
            providesAnalytics: true,
          }),
          UiKitsModule,
          CommonModule,
          ReactiveFormsModule,
          IonicModule,
          UiKitsModule,
        ],
        providers: [
          {
            provide: PaymentService,
            useValue: {
              getPaymentHistory: jest.fn().mockReturnValue(of({})),

              getWallets: jest.fn().mockReturnValue(of({})),
              refreshPayments: jest.fn(),
              refreshHistory: jest.fn(),
              getUpcomingPayments: jest.fn().mockReturnValue(of([])),
              refreshAll: jest.fn(),
            },
          },
          {
            provide: MakePaymentService,
            useValue: {
              setActivePayment: jest.fn(),
              getActivePayment: jest.fn().mockReturnValue(mockPaymentData),
              getReturnPathFromAmountMethodPages: jest.fn().mockReturnValue('path.to.index'),
            },
          },
        ],
      }).compileComponents();

      makePaymentService = TestBed.inject(MakePaymentService);
      routerService = TestBed.inject(RouterService);
      analyticService = TestBed.inject(AnalyticsService);
      alertControllerMock = TestBed.inject(AlertController) as any;
      makePaymentService.setActivePayment({} as UpcomingPayment); // Todo:: use valid mock

      fixture = TestBed.createComponent(PaymentSelectAmountPage);
      component = fixture.componentInstance;
      fixture.detectChanges();
      component.ionViewWillEnter();
      await fixture.whenStable();
      fixture.detectChanges();
      await fixture.whenRenderingDone();
    } catch (error) {
      console.error(error);
      fail(error.message);
    }
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match screenshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should select the minimum due', () => {
    const minimumDue = fixture.debugElement.query(By.css('#item-minimum'));
    minimumDue.triggerEventHandler('click', null);
    fixture.detectChanges();
    const doneButton = fixture.debugElement.query(By.cssAndText('ion-button', 'Done'));
    doneButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(makePaymentService.setActivePayment).toHaveBeenCalled();
    expect(routerService.back).toHaveBeenCalled();
  });

  it('should select the remaining due', () => {
    const remainingDue = fixture.debugElement.query(By.css('#item-remaining'));
    remainingDue.triggerEventHandler('click', null);
    fixture.detectChanges();
    const doneButton = fixture.debugElement.query(By.cssAndText('ion-button', 'Done'));
    doneButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(makePaymentService.setActivePayment).toHaveBeenCalled();
    expect(routerService.back).toHaveBeenCalled();
  });

  it('should select the other amount and insert a input', () => {
    const otherAmountInput = fixture.nativeElement.querySelector('[formcontrolname= otherAmount]');
    otherAmountInput.value = '123.00';
    otherAmountInput.dispatchEvent(new Event('ionChange'));
    fixture.detectChanges();
    const doneButton = fixture.debugElement.query(By.cssAndText('ion-button', 'Done'));
    doneButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(makePaymentService.setActivePayment).toHaveBeenCalled();
    expect(routerService.back).toHaveBeenCalled();
  });

  it('should dispaly alert informing user it cant pay over the minimum amount', () => {
    const otherAmountInput = fixture.nativeElement.querySelector('[formcontrolname= otherAmount]');
    otherAmountInput.value = '11.00';
    otherAmountInput.dispatchEvent(new Event('ionChange'));
    fixture.detectChanges();
    const doneButton = fixture.debugElement.query(By.cssAndText('ion-button', 'Done'));
    doneButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(alertControllerMock.create).toHaveBeenCalledWith(
      expect.objectContaining({
        header: 'You are paying less than the minimum due.',
        message: 'Paying less than the minimum may result in policy cancelling.',
      })
    );
  });

  it('should dispaly alert informing user it cant pay over the remaining amount', () => {
    const otherAmountInput = fixture.nativeElement.querySelector('[formcontrolname= otherAmount]');
    otherAmountInput.value = '11000.00';
    otherAmountInput.dispatchEvent(new Event('ionChange'));
    fixture.detectChanges();
    const doneButton = fixture.debugElement.query(By.cssAndText('ion-button', 'Done'));
    doneButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(alertControllerMock.create).toHaveBeenCalledWith(
      expect.objectContaining({
        header: 'Payment error',
        message: 'You cannot pay more than the remaining premium.',
      })
    );
  });

  it('should track the cancel button', () => {
    const cancelButton = fixture.debugElement.query(By.cssAndText('ion-button', 'Cancel'));
    cancelButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(analyticService.trackEvent).toHaveBeenCalledWith(
      EventName.PAYMENT_AMOUNT_CANCELLED,
      Category.payments
    );
  });
});
