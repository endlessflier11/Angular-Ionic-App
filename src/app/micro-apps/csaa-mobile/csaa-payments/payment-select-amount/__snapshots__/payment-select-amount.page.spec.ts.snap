// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentSelectAmountPage should match screenshot 1`] = `
<app-csaa-payment-select-amount
  __ngContext__={[Function LRootView]}
  alertCtrl={[Function AlertControllerMock]}
  amountForm={[Function FormGroup]}
  analyticsService={[Function AnalyticsMockService]}
  configService={[Function ConfigMockService]}
  currentTheme=""
  formBuilder={[Function FormBuilder]}
  isIos="false"
  makePaymentService={[Function Object]}
  paymentData={[Function Object]}
  payments={[Function Array]}
  platform={[Function Platform]}
  routerService={[Function RouterMockService]}
  store={[Function Store]}
>
  <ion-header
    class="csaa"
  >
    <ion-toolbar>
      <ion-buttons
        slot="start"
      >
        <ion-button>
          Cancel
        </ion-button>
      </ion-buttons>
      <ion-title>
        Amount
      </ion-title>
      <ion-buttons
        slot="end"
      >
        <ion-button>
          Done
        </ion-button>
      </ion-buttons>
    </ion-toolbar>
  </ion-header><ion-content
    class="csaa-item-list csaa"
  >
    <h3
      class="ion-text-center"
    >
      Select an amount
    </h3>
    <form
      class="ng-untouched ng-pristine ng-valid"
      novalidate=""
    >
      <ion-list
        lines="full"
      >
        <ion-item
          id="item-minimum"
        >
          <ion-label>
            Minimum Due
          </ion-label>
          <span
            slot="end"
          >
             $20.00 
          </span>
        </ion-item>
        <ion-item
          id="item-remaining"
        >
          <ion-label>
            Remaining Premium
          </ion-label>
          <span
            item-end=""
          >
             $10,000.00 
          </span>
        </ion-item>
        <ion-item
          class="active"
          id="item-other"
        >
          <ion-label>
            Other Amount
          </ion-label>
          <ion-input
            class="ion-text-right ng-untouched ng-pristine ng-valid"
            formcontrolname="otherAmount"
          />
        </ion-item>
      </ion-list>
    </form>
  </ion-content>
</app-csaa-payment-select-amount>
`;
